state_machine/applications/system-check/tests/imu_test.py

```diff
@@ -0,0 +1,270 @@
+        gyro_total += numpy.array(cubesat.gyro())
+        time.sleep(wait_time / 10)
+
+    # collect magnetometer reading at the end of wait time
+    end_mag = numpy.array(cubesat.magnetic())
+
+    # calculate the average acc and gyro readings
+    acc_final = acc_total / 10
+    gyro_final = gyro_total / 10
+
+    # calculate the change in the total magnetometer reading
+    mag = norm(end_mag - start_mag)
+
+    # return acc, gyro, change in mag
+    print("Data Collection Complete")
+    return acc_final, gyro_final, mag
+
+
+def stationary_imu_test(result_dict):
+    """
+    Return acceleration, gyro, and magnetic readings from request_imu_data
+    If None, update result dictionary that test was not run
+    Else, check that the norms of the average accelerometer and gyroscope
+    readings are near 0 and update result dictionary
+    """
+    # prompt user and get imu data
+    prompt = "Please leave the cubesat stationary on a table.\n"
+    acc, gyro, mag = request_imu_data(prompt)
+
+    if acc is None and gyro is None:
+        # user did not enter "y", cancel the test
+        result_dict["IMU_AccStationary"] = (
+            "Stationary test not completed.", None)
+        result_dict["IMU_GyroStationary"] = (
+            "Stationary test not completed.", None)
+        return result_dict
+
+    # else continue running the test
+    acc_string = f"Acc: {tuple(acc)} (m/s^2)"
+    gyro_string = f"Gyro: {tuple(gyro)} (deg/s)"
+
+    # if total acceleration ~= 9.8 m/s^2 and gyro ~= 0 deg/s, true
+    acc_is_stationary = 9.8 - norm(acc) < 0.5
+    gyro_is_stationary = norm(gyro) < 0.5
+
+    # print result to user
+    if acc_is_stationary:
+        print("Stationary acceleration is near g m/s^2.")
+    else:
+        print("Stationary acceleration is not near g m/s^2.")
+
+    if gyro_is_stationary:
+        print("Stationary gyroscope reading is near 0 deg/s.")
+    else:
+        print("Stationary gyroscope reading is not near 0 deg/s.")
+
+    # update result dictionary
+    result_dict["IMU_AccStationary"] = (acc_string, acc_is_stationary)
+    result_dict["IMU_GyroStationary"] = (gyro_string, gyro_is_stationary)
+    return result_dict
+
+
+def moving_imu_test(result_dict):
+    """
+    Return acceleration, gyro, and magnetic readings from request_imu_data
+    If None, update result dictionary that test was not run
+    Else, check that the norm of the average acceleration reading is positive
+    and update result dictionary
+    """
+    # record acceleration
+    prompt = f"""Please move the cubesat around for {wait_time} seconds once
+you start the test.\n"""
+    acc, gyro, mag = request_imu_data(prompt)
+
+    if acc is None:
+        # user entered "n", cancel the test
+        result_dict["IMU_AccMoving"] = ("Moving test not completed.", None)
+        return result_dict
+
+    # else continue running the test
+    acc_string = (f"Acc: {tuple(acc)} (m/s^2)")
+    acc_is_moving = norm(acc) >= 9.8
```

Comment:
I'm not sure that this is a good threshold, because we can technically cancel out acceleration by gravity if we have a significant amount of upward acceleration. I've tried the move test multiple times, and it tends to fail on this threshold when I move the mainboard up and down, and pass when I move it side to side.

At the same time, testing with a threshold of 0.5 m/s^2 will mean that we can pass a moving test even when the cubesat is stationary. 

I'd love to hear your thoughts @jbwillis @thetazero

Reply:
If this is still an issue, I really recommend returning the array of readings rather than the average of the readings.
And then doing the 
```python
delta_a = norm(prev-cur)
``` 
over the array, giving you the change in acceleration at every time step.
If this is sufficiently large you can consider it to be moving. 
Let me know if this works for you

Reply:
@thetazero is on the right track. This is basically taking the derivative of the acceleration. 

I'd say it's good enough to change your instructions to have the person accelerate it "side to side". This is really only meant to test that we're getting reasonable readings from the accelerometer. Doesn't need to be complicated or super accurate.

Another idea is to have the person hold it in three different orientations and then measure the gravity vector in those orientations. Maybe something like +X up, +Y up, +Z up. In this case. you can assume that the gravity vector is approximately `[-9.8, 0, 0], [0, -9.8, 0], [0, 0, -9.8]` and check for that. Though that will also require knowing the orientation of the IMU with respect to the assembled satellite (which we can figure out).

Reply:
I do like the idea of approximating the derivative for each function. However, we still don't have any measure of what "sufficiently large" is. I can test for a change in at least 0.5 to 1 m/s^2? 

I think the most reasonable solution is to either just tell the user to accelerate it side to side (which is prone to human error) or to do the gravity vector test in different orientations. The gravity vector idea is probably the most fool proof. I can also write it so the user doesn't have to know what the orientation of the IMU is. This is one way we can implement it:

First case/side: check that any one of the vectors is near 9.8 and the others are near 0
Second case/side: check that one of the remaining two vectors is near 9.8 and the others are near 0
Third case/side: check that the remaining vector is near 9.8 and the others are near 0